src/compiler/parser.ts

```diff
@@ -5227,18 +5265,38 @@ namespace ts {
 
         // DECLARATIONS
 
-        function parseArrayBindingElement(): ArrayBindingElement {
+        function parseArrayBindingElementInSpeculation(): ArrayBindingElement | Fail {
+            return parseArrayBindingElementWorker(/*inSpeculation*/ true);
+        }
+        function parseArrayBindingElementNoSpeculation(): ArrayBindingElement {
+            return parseArrayBindingElementWorker(/*inSpeculation*/ false) as ArrayBindingElement;
+        }
+        function parseArrayBindingElementWorker(inSpeculation: boolean): ArrayBindingElement | Fail {
             if (token() === SyntaxKind.CommaToken) {
                 return <OmittedExpression>createNode(SyntaxKind.OmittedExpression);
             }
             const node = <BindingElement>createNode(SyntaxKind.BindingElement);
             node.dotDotDotToken = parseOptionalToken(SyntaxKind.DotDotDotToken);
-            node.name = parseIdentifierOrPattern();
-            node.initializer = parseInitializer(/*inParameter*/ false);
+            const name = parseIdentifierOrPattern(inSpeculation);
+            if (isFail(name)) {
+                return Fail;
+            }
+            node.name = name;
+            const init = parseInitializer(/*inParameter*/ false, inSpeculation);
+            if (isFail(init)) {
+                return Fail;
+            }
+            node.initializer = init;
             return finishNode(node);
         }
 
-        function parseObjectBindingElement(): BindingElement {
+        function parseObjectBindingElementInSpeculation(): BindingElement | Fail {

```

Comment:
As above, I'd recommend `tryParseObjectBindingElement` and `parseObjectBindingElement`.